### PR TITLE
Add support for importing existing resources.

### LINKED ISSUE #1662: Resource Adoption
While pondering https://github.com/pulumi/pulumi/pull/1654, we've realized that there is an interesting operation lurking that we have tentatively called "adoption". where a user expresses in their Pulumi program that they are "adopting" a resource that was previously *not* managed by Pulumi into their program so that, in the future, it *is* managed by Pulumi.

It is an interesting and useful scenario to be able to "adopt" a resource without having to delete it. This can't be expressed in Pulumi today, because:

1. The user needs to express that a resource should be created and request that it have a specific ID - the same ID of the existing resource that a user intends to adopt.
2. The engine, when it sees an adoption where the requested ID is equivalent to the read ID, needs to not delete the resource.

This implies some changes to the language host and language projection layers of Pulumi. We should think about whether we want these changes and what they would look like.

### PR DESCRIPTION
**Pull request by pgavlin**
A resource can be imported by setting the `import` property in the
resource options bag when instantiating a resource. In order to
successfully import a resource, its desired configuration (i.e. its
inputs) must not differ from its actual configuration (i.e. its state)
as calculated by the resource's provider.

There are a few interesting state transitions hiding here when importing
a resource:
1. No prior resource exists in the checkpoint file. In this case, the
   resource is simply imported.
2. An external resource exists in the checkpoint file. In this case, the
   resource is imported and the old external state is discarded.
3. A non-external resource exists in the checkpoint file and its ID is
   different from the ID to import. In this case, the new resource is
   imported and the old resource is deleted.
4. A non-external resource exists in the checkpoint file, but the ID is
   the same as the ID to import. In this case, the import ID is ignored
   and the resource is treated as it would be in all cases except for
   changes that would replace the resource. In that case, the step
   generator issues an error that indicates that the import ID should be
   removed: were we to move forward with the replace, the new state of
   the stack would fall under case (3), which is almost certainly not
   what the user intends.

Fixes #1662.

### CODE REVIEW

**Comment by pgavlin** [comment]
We could choose to make (4) an error in all situations, but this would require that the user immediately remove all import declarations from their program after importing a resource.

**Comment by pgavlin** [comment]
We'll need to think about how we want to support import in dynamic providers: the implementation of `ImportStep` intentionally does not pass inputs or state in its call to `Read`, so the dynamic provider has no program to pass to `requireFromString`. It would be pretty straightforward to make this work with first-class dynamic providers or to hack things up so that inputs are passed to `Read` during an import specifically for dynamic resources. It is possible that we may be able to always pass the inputs to `Read`, but I avoided doing so to maintain the current situation where inputs are only passed to `Read` during a `Refresh`.

**Comment by ellismg** [comment]
> It is possible that we may be able to always pass the inputs to Read, but I avoided doing so to maintain the current situation where inputs are only passed to Read during a Refresh.

Do you know why this decision was made to not pass inputs to Read in general?

**Comment by pgavlin** [comment]
> Do you know why this decision was made to not pass inputs to Read in general?

We don't have inputs for `Read` when it is called via `.get`: instead, the parameters passed in by the program are treated as state.

We may want to revisit that decision, but I think we should do so independent of this PR.

File: pkg/backend/display/rows.go [hunk 1]
```diff
@@ -348,14 +348,21 @@ func (data *resourceRowData) ColorizedColumns() []string {
...
 		for _, outputStep := range data.outputSteps {
 			if outputStep.Op == deploy.OpReplace {
 				step = outputStep
 			}
 		}
+
+	case deploy.OpImport:
+		// If we're doing an import, see if we have the imported state to diff.
+		if len(data.outputSteps) != 0 {
+			step = data.outputSteps[0]
```
Comment (ellismg) [inline pkg/backend/display/rows.go]: I assume that there's some hidden invariant here that `data.outputSteps[0]` is correct.  Should we be asserting something about `step.Op` at this point?  Would it make more sense to write the above as a range like we have for Create/Replace?

File: pkg/resource/deploy/step.go [hunk 2]
```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
...
+			s.new.InitErrors = initErr.Reasons
+		} else {
+			return rst, nil, err
+		}
+	}
+	if read.Outputs == nil {
+		return rst, nil, errors.Errorf("resource '%v' does not exist", s.new.ID)
+	}
+	if read.Inputs == nil {
+		return resource.StatusOK, nil, errors.Errorf(
```
Comment (ellismg) [inline pkg/resource/deploy/step.go]: Just wondering - do our existing released plugins today fail in this way, or is it just much older versions?

File: pkg/resource/deploy/step.go [hunk 3]
```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
...
+			"provider does not support importing resources; please try updating the '%v' plugin",
+			s.new.URN.Type().Package())
+	}
+	s.new.Outputs = read.Outputs
+
+	// Magic up an old state so the frontend can display a proper diff.
```
Comment (ellismg) [inline pkg/resource/deploy/step.go]: Might be nice to explain the magic you are doing here, so it was clear why we choose the values we did.  I think the state you are creating here is the logical state that would be returned by a Read step, and I think that's the correct thing to do, but I am not sure.

File: pkg/resource/deploy/step.go [hunk 4]
```diff
@@ -709,6 +709,135 @@ func (s *RefreshStep) Apply(preview bool) (resource.Status, StepCompleteFunc, er
...
+
+	// Diff the user inputs against the provider inputs. If there are any differences, fail the import.
+	diff, err := diffResource(s.new.URN, s.new.ID, s.old.Inputs, s.old.Outputs, s.new.Inputs, prov, preview)
+	if err != nil {
+		return rst, nil, err
+	}
+	s.diffs, s.detailedDiff = diff.ChangedKeys, diff.DetailedDiff
+
+	if diff.Changes != plugin.DiffNone {
+		const message = "inputs to import do not match the existing resource"
```
Comment (ellismg) [inline pkg/resource/deploy/step.go]: I feel like this message is not great, but I honestly don't have a better suggestion.

File: pkg/resource/deploy/step_generator.go [hunk 5]
```diff
@@ -319,6 +334,18 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 		// If there were changes, check for a replacement vs. an in-place update.
 		if diff.Changes == plugin.DiffSome {
 			if diff.Replace() {
+				// If the goal state specified an ID, issue an error: the replacement will change the ID, and is
+				// therefore incompativle with the goal state.
```
Comment (ellismg) [inline pkg/resource/deploy/step_generator.go]: NIT: "incompatible"

File: sdk/proto/provider.proto [hunk 6]
```diff
@@ -196,7 +196,7 @@ message ReadRequest {
 message ReadResponse {
     string id = 1;                         // the ID of the resource read back (or empty if missing).
     google.protobuf.Struct properties = 2; // the state of the resource read from the live environment.
-    google.protobuf.Struct inputs = 3;     // the inputs for this resource that would be returned from Check.
+	google.protobuf.Struct inputs = 3;     // the inputs for this resource that would be returned from Check.
```
Comment (ellismg) [inline sdk/proto/provider.proto]: NIT: Hard tabs

File: sdk/proto/provider.proto [hunk 7]
```diff
@@ -224,5 +224,5 @@ message ErrorResourceInitFailed {
     string id = 1;                         // the ID of the created resource.
     google.protobuf.Struct properties = 2; // any properties that were computed during updating.
     repeated string reasons = 3;           // error messages associated with initialization failure.
-    google.protobuf.Struct inputs = 4;     // the current inputs to this resource (only applicable for Read)
+	google.protobuf.Struct inputs = 4;     // the current inputs to this resource (only applicable for Read)
```
Comment (ellismg) [inline sdk/proto/provider.proto]: NIT: Hard tabs

File: sdk/proto/resource.proto [hunk 8]
```diff
@@ -88,6 +88,7 @@ message RegisterResourceRequest {
     bool acceptSecrets = 13;           // when true operations should return secrets as strongly typed.
     repeated string additionalSecretOutputs = 14;  // a list of output properties that should also be treated as secret, in addition to ones we detect.
     repeated string aliases = 15;      // a list of additional URNs that shoud be considered the same.
+	string importId = 16;              // if set, this resource's state should be imported from the given ID.
```
Comment (ellismg) [inline sdk/proto/resource.proto]: NIT: Hard tabs

File: tests/integration/read/import_acquire/import_acquire_test.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import (
+	"testing"
+
+	"github.com/pulumi/pulumi/pkg/testing/integration"
+)
+
+// Test that the engine is capable of assuming control of a resource that was external.
+func TestImportAcquire(t *testing.T) {
+	t.Skipf("import does not yet work with dynamic providers")
```
Comment (ellismg) [inline tests/integration/read/import_acquire/import_acquire_test.go]: Could we add an issue for this?

File: sdk/python/lib/test/langhost/util.py [hunk 10]
```diff
@@ -94,16 +94,17 @@ def RegisterResource(self, request, context):
...
             property_dependencies[key] = sorted(list(value.urns))
 
         outs = {}
         if type_ != "pulumi:pulumi:Stack":
-            outs = self.langhost_test.register_resource(
-                context, self.dryrun, type_, name, props, deps, parent, custom, protect, provider,
-                property_dependencies, delete_before_replace, ignore_changes, version)
+            rrsig = signature(self.langhost_test.register_resource)
+            args = [context, self.dryrun, type_, name, props, deps, parent, custom, protect, provider, property_dependencies, delete_before_replace, ignore_changes, version, import_]
+            outs = self.langhost_test.register_resource(*args[0:len(rrsig.parameters)])
```
Comment (ellismg) [inline sdk/python/lib/test/langhost/util.py]: Just wondering what caused you to make these changes (vs just appending yet another argument)

File: sdk/python/lib/pulumi/resource.py [hunk 11]
```diff
@@ -84,6 +84,13 @@ class ResourceOptions:
     An optional existing ID to load, rather than create.
     """
 
+    import_: Optional[str]
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/python/lib/pulumi/resource.py]: You are correct with respect to the chosen name.

I had originally used `importId` in NodeJS and `import_id` here but changed to `import` and `import_` because I had received some early feedback that the first two choices didn't read well. I don't have terribly strong feelings here, so I am happy to roll with whatever the majority feels is best.

File: pkg/resource/deploy/step.go (hunk 2 above)
Comment (pgavlin) [reply pkg/resource/deploy/step.go]: Only older versions will fail in this fashion. FWIW, I considered adding something to `ConfigureResponse` to indicate that a provider does support importing resources, which would allow a provider to return a nil input bag here if it can properly support `Check` and `Diff` without old inputs.

File: pkg/resource/deploy/step.go (hunk 4 above)
Comment (pgavlin) [reply pkg/resource/deploy/step.go]: Yeah, same.

File: pkg/resource/deploy/step.go (hunk 3 above)
Comment (pgavlin) [reply pkg/resource/deploy/step.go]: Yes, that's right. I will expand the comment.

File: sdk/python/lib/test/langhost/util.py (hunk 10 above)
Comment (pgavlin) [reply sdk/python/lib/test/langhost/util.py]: > Just wondering what caused you to make these changes (vs just appending yet another argument)

Updating every single Python test each time we make a change is both tedious and inconsistent with the flow for Node.

File: sdk/python/lib/pulumi/resource.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply sdk/python/lib/pulumi/resource.py]: I would be happy to keep `import` (the nice name) in NodeJS and `import_id` in Python.  These are going to differ anyway, and I think that `import_id` reads much better than `import_` FWIW.

**Comment by stack72** [comment]
So I’m super interested in why import isn’t a specific company as like it is in TF? I mean we could do something like “pulumi import id stackname”

**Comment by lukehoban** [comment]
> So I’m super interested in why import isn’t a specific company as like it is in TF? I mean we could do something like “pulumi import id stackname”

@pgavlin Could you summarize the overall plan here in https://github.com/pulumi/pulumi/issues/1635?  I believe that issue is tracking adding an explicit `pulumi import`, and is planned as a follow-up to this work.  But would be great to summarize there what steps we plan to take and what scenarios will be supported by each.

**Comment by pgavlin** [comment]
For an example of the flow here, see https://asciinema.org/a/Ov7s3Xxa0YtXzWwVtleyGYfCw, which imports a k8s service and deployment.

File: sdk/python/lib/pulumi/resource.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/python/lib/pulumi/resource.py]: Using a trailing underscore to avoid conflicts with language keywords is PEP 8 style, FWIW: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles (look for "`single_trailing_underscore_`: used by convention to avoid conflicts with Python keyword").

File: sdk/proto/resource.proto (hunk 8 above)
Comment (pgavlin) [reply sdk/proto/resource.proto]: Fixed.

File: sdk/proto/provider.proto (hunk 7 above)
Comment (pgavlin) [reply sdk/proto/provider.proto]: Fixed.

File: sdk/proto/provider.proto (hunk 6 above)
Comment (pgavlin) [reply sdk/proto/provider.proto]: Fixed.

File: pkg/resource/deploy/step_generator.go (hunk 5 above)
Comment (pgavlin) [reply pkg/resource/deploy/step_generator.go]: Fixed.

File: pkg/backend/display/rows.go (hunk 1 above)
Comment (pgavlin) [reply pkg/backend/display/rows.go]: Fixed.

**Comment by pgavlin** [comment]
Ping. I'd love to get this landed ASAP.

**Comment by stack72** [comment]
So the code looks good - it looks like we have covered the bases. The flow to me still feels strange... but only because I always imagined this was a CLI based thing like it is in Terraform. Just to confirm the scenario from the video:

* we deployed redis using kubectl
* we created a basic manifest that just pointed us at the *basic* service definition for the redis deployment and asked for it to be imported
* pulumi up will do the import and show us the differences via `details`
* we take all of the code from the diff and add it to our .ts (or whatever) files
* then we can run pulumi up and nothing will change

This feels like a similar flow to what Terraform does as well - so that fits. The big question for me is, how do we treat an import that has the basic skeleton of the code and the user runs `pulumi up --yes` will it trample all over their deployment? If so, are there plans to gate that from being able to be done?

**Comment by pgavlin** [comment]
> The big question for me is, how do we treat an import that has the basic skeleton of the code and the user runs pulumi up --yes will it trample all over their deployment? If so, are there plans to gate that from being able to be done?

The `pulumi up` will fail if there are any differences detected between the desired state of a resource to import as described by the program and the actual state of the resource. So if I understand your question correctly, this is already gated.

**Comment by stack72** [comment]
> > The big question for me is, how do we treat an import that has the basic skeleton of the code and the user runs pulumi up --yes will it trample all over their deployment? If so, are there plans to gate that from being able to be done?
> 
> The `pulumi up` will fail if there are any differences detected between the desired state of a resource to import as described by the program and the actual state of the resource. So if I understand your question correctly, this is already gated.

Excellent so the flow feels correct and I like it - LGTM!

**Comment by CyrusNajmabadi** [comment]
I'm going to look later tonight.  Thanks.

File: pkg/resource/deploy/step.go (hunk 4 above)
Comment (CyrusNajmabadi) [reply pkg/resource/deploy/step.go]: could we print out at least some of things that are different?

File: pkg/resource/deploy/step.go [hunk 12]
```diff
@@ -808,6 +943,10 @@ func (op StepOp) RawPrefix() string {
 		return "< "
 	case OpDiscardReplaced:
 		return "<<"
+	case OpImport:
+		return "= "
```
Comment (CyrusNajmabadi) [inline pkg/resource/deploy/step.go]: any thought on this?  as opposed to maybe `<=` (to indicate we're sucking it in)?
(yes, this is bikeshedding to the nth degree).

Comment (pgavlin) [reply pkg/resource/deploy/step.go]: Honestly I’d like to use the `>` symbol that’s currently allocated to `Read`, but I prefer `=` otherwise.

**Comment by CyrusNajmabadi** [comment]
LGTM.  Nothing striking me as seriously problematic.

File: pkg/resource/deploy/step.go (hunk 4 above)
Comment (pgavlin) [reply pkg/resource/deploy/step.go]: We do that as part of the usual progress/diff view.

Comment (CyrusNajmabadi) [reply pkg/resource/deploy/step.go]: Great.  Thanks!

**Comment by jeid64** [comment]
Howdy, I built this PR on my own laptop and am trying to import some pre-existing AWS resources into Pulumi using Python.

When I try to pulumi up and import a pre-existing S3 bucket I get this message

`Previewing update (dev):

     Type                      Name                  Plan       Info
     pulumi:pulumi:Stack       aws-python-dev
 =   └─ aws:ec2:SecurityGroup  preprod-testjitsua-1  import     [diff: ~name]; 1 warning

Diagnostics:
  aws:ec2:SecurityGroup (preprod-testjitsua-1):
    warning: inputs to import do not match the existing resource; importing this resource will fail

Resources:
    = 1 to import
    1 unchanged

Do you want to perform this update? details
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::aws-python::pulumi:pulumi:Stack::aws-python-dev]
    = aws:ec2/securityGroup:SecurityGroup: (import)
        [id=sg-4c41ba2b]
        [urn=urn:pulumi:dev::aws-python::aws:ec2/securityGroup:SecurityGroup::preprod-testjitsua-1]
      ~ name: "preprod-testjitsua-1" => "preprod-testjitsua-1-be29ea6"

Do you want to perform this update?
  yes
> no`

It seems to always be appending a random string at the end of my SG name and I'm unable to import it? Am I setting up my resource opts incorrectly? I set the import_ variable to the name of my SG

`resource_opts = resource.ResourceOptions(import_="sg-4c41ba2b")
tags = { "Environment": "preprod", "Name": "preprod-testjitsua-1", "ServiceName": "testjitsua-1"}
security_group = ec2.SecurityGroup('preprod-testjitsua-1', resource_opts, description="Jiujitsu Test Servers", tags=tags)`

@pgavlin

**Comment by lukehoban** [comment]
@jeid64 I would expect you need to specify the `name` of the security group:

```py
security_group = ec2.SecurityGroup('preprod-testjitsua-1', resource_opts, name="preprod-testjitsua-1", description="Jiujitsu Test Servers", tags=tags)
```

Note that the first argument to the `SecurityGroup` is the name you want Pulumi to use to track this, not the name of the security group itself in AWS.  So these could be different if needed.

**Comment by jeid64** [comment]
That was it, thank you so much! Playing around with import now.
